src/complete.cpp

```diff
@@ -838,19 +836,24 @@ static bool short_ok(const wcstring &arg, const complete_entry_opt_t *entry,
     return result;
 }
 
+struct cmd_data_t {
+    const wcstring &name;
+    bool has_path;
+};
+
 /// Load command-specific completions for the specified command.
-static void complete_load(const wcstring &name, bool reload) {
-    // We have to load this as a function, since it may define a --wraps or signature.
-    // See issue #2466.
-    function_load(name);
-    completion_autoloader.load(name, reload);
+static void complete_load(const struct cmd_data_t *cmd, bool reload) {
+    // don't bother to autoload if there's no associated command able to use potential completions
+    // (#2365)
+    if (cmd->has_path || builtin_exists(cmd->name) || function_exists(cmd->name)) {
+        completion_autoloader.load(cmd->name, reload);
+    }
 }
 
 /// Performed on main thread, from background thread. Return type is ignored.
-static int complete_load_no_reload(wcstring *name) {
-    assert(name != NULL);
+static int complete_load_no_reload(const struct cmd_data_t *cmd) {
```

Comment:
This cmd_data_t should remain a pointer (since it's bouncing across threads), so this is fine.